src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => delete newState[key]);
+        break;
+      case 'clear':
+        newState = {};
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    stateHistory.push(newState);
+    currentState = newState;
+  });
+
+  return stateHistory;
```

Comment:
The function `transformStateWithClones` is well-structured and follows the task requirements. However, there is a small improvement that can be made: the assignment of `newState` to `currentState` is performed at the end of each `forEach` iteration (line 31). This action is repeated regardless of the `switch` case, and therefore can be moved outside the `switch` statement to adhere to the DRY (Don't Repeat Yourself) principle. This would slightly clean up the code.